parsec/recursive.h

```diff
@@ -12,42 +12,44 @@
 #include "parsec/mca/device/device.h"
 #include "parsec/data_dist/matrix/matrix.h"
 
+typedef struct parsec_recursive_cb_data_s parsec_recursive_cb_data_t;
+typedef void (*parsec_recursive_callback)(parsec_taskpool_t*, const parsec_recursive_cb_data_t* );
+
 typedef struct parsec_recursive_cb_data_s {
     parsec_execution_stream_t    *es;
-    parsec_task_t *context;
-    void (*destruct)( parsec_taskpool_t * );
+    parsec_task_t                *task;
+    parsec_recursive_callback     callback;
     int nbdesc;
-    parsec_data_collection_t *desc[1];
+    parsec_data_collection_t      *desc[1];
 } parsec_recursive_cb_data_t;
 
 static inline int parsec_recursivecall_callback(parsec_taskpool_t* tp, void* cb_data)
 {
     int i, rc = 0;
     parsec_recursive_cb_data_t* data = (parsec_recursive_cb_data_t*)cb_data;
 
-    rc = __parsec_complete_execution(data->es, data->context);
+    rc = __parsec_complete_execution(data->es, data->task);
 
-    for(i=0; i<data->nbdesc; i++){
+    for( i = 0; i < data->nbdesc; i++ ) {
         parsec_tiled_matrix_destroy( (parsec_tiled_matrix_t*)(data->desc[i]) );
         free( data->desc[i] );
     }
 
-    data->destruct( tp );
+    data->callback( tp, data );
```

Comment:
This must be called before `__parsec_complete_execution` or you can get some nasty races where `data->task` is no longer valid inside of `data->callback`.

Reply:
I cannot find the patch I sent you for this a while back. Do you have it around ?

Reply:
You can cherry-pick 1e15cdd3c2c82e5e4b32d5c85e44936442df67fe from my fork.